lib/experimental/script-modules.php

```diff
@@ -197,3 +197,94 @@ function gutenberg_dequeue_module( $module_identifier ) {
 	_deprecated_function( __FUNCTION__, 'Gutenberg 17.6.0', 'wp_dequeue_script_module' );
 	wp_script_modules()->dequeue( $module_identifier );
 }
+
+
+/**
+ * Print data associated with Script Modules in Script tags.
+ *
+ * This embeds data in the page HTML so that it is available on page load.
+ *
+ * Data can be associated with a given Script Module by using the
+ * `scriptmoduledata_{$module_id}` filter.
+ *
+ * The data for a given Script Module will be JSON serialized in a script tag with an ID
+ * like `wp-scriptmodule-data_{$module_id}`.
+ */
+function gutenberg_print_script_module_data(): void {
+	$get_marked_for_enqueue = new ReflectionMethod( 'WP_Script_Modules', 'get_marked_for_enqueue' );
+	$get_import_map         = new ReflectionMethod( 'WP_Script_Modules', 'get_import_map' );
+
+	$modules = array();
+	foreach ( array_keys( $get_marked_for_enqueue->invoke( wp_script_modules() ) ) as $id ) {
+		$modules[ $id ] = true;
+	}
+	foreach ( array_keys( $get_import_map->invoke( wp_script_modules() )['imports'] ) as $id ) {
+		$modules[ $id ] = true;
+	}
+
+	foreach ( array_keys( $modules ) as $module_id ) {
+		/**
+		 * Filters data associated with a given Script Module.
+		 *
+		 * Script Modules may require data that is required for initialization or is essential to
+		 * have immediately available on page load. These are suitable use cases for this data.
+		 *
+		 * This is best suited to a minimal set of data and is not intended to replace the REST API.
+		 *
+		 * If the filter returns no data (an empty array), nothing will be embedded in the page.
+		 *
+		 * The data for a given Script Module, if provided, will be JSON serialized in a script tag
+		 * with an ID like `wp-scriptmodule-data_{$module_id}`.
+		 *
+		 * The dynamic portion of the hook name, `$module_id`, refers to the Script Module ID that
+		 * the data is associated with.
+		 *
+		 * @param array $data The data that should be associated with the array.
+		 */
+		$data = apply_filters( "scriptmoduledata_{$module_id}", array() );
+
+		if ( ! empty( $data ) ) {
```

Comment:
Should we check that `$data` is indeed an array in addition to checking
if it is not empty?

Users could e.g. accidentally return a string from the filter in which case whatever's
printed in the script tag won't be parsed correctly on the frontend.

Reply:
I've gone back and forth, but I'll add an `is_array` check. My thoughts below.

---

**The reason data needs to be an array is filters**, not the frontend. Filters need have an interface they can work with. Filters don't need to know about each other, and different filters can add (or modify) data theoretically without interfering.

There's no technical reason the frontend needs an object, if a module wanted to use a string or other data type that would be fine. The limitation is really with the PHP filters.

That brings up another interesting point, filters could look like this:

```php
function filter( $data ) {
  $data[] = 'x';
  return $data;
}
```

In which case we'd end up with an array instead of an object on the frontend.

But what if `$data` isn't an array? Then the `empty` check becomes messy because we can't pass falsy values 😕 .

It also seems like a bad practice to let some modules get away way not using an array, because then they violate the filter interface and may cause errors with other filters.